core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -205,4 +238,24 @@ private[fs2] object Algebra {
     }
     fr.translate[Algebra[G,O,?]](algFtoG)
   }
+
+  /**
+    * Specific version of `Free.asHandler` that will output any CloseScope that may have
+    * been opened before the `rsn` was yielded
+    */
+  def asHandler[F[_], O](free: FreeC[Algebra[F,O,?],Unit], e: Throwable): FreeC[Algebra[F,O,?],Unit] = {
+    free.viewL.get match {
+      case FreeC.Pure(_) => FreeC.Fail(e)
+      case FreeC.Fail(e2) => FreeC.Fail(e)
+      case bound: FreeC.Bind[Algebra[F,O,?], _, Unit] =>
+        val f = bound.f.asInstanceOf[
+          Either[Throwable,Any] => FreeC[Algebra[F,O,?], Unit]]
+        val fx = bound.fx.asInstanceOf[FreeC.Eval[Algebra[F,O,?],_]].fr
+        fx match {
+          case _: Algebra.CloseScope[F,_] => FreeC.Bind(FreeC.Eval(fx), { (_: Either[Throwable, Any]) => f(Left(e)) })
```

Comment:
What if there's another `CloseScope` in the result of `f(fx)`?

Reply:
Not sure if we need deep search this. I am quite certain we always get the closest to root scope that may be needed to close when flatMap fails. I was not able to come with stream structure where that would not be enough.

Reply:
Why do we only care about the closest `CloseScope`?